### PR TITLE
Ensure that leading/trailing code does not overlap with student code

### PR DESCRIPTION
**Pull request by jgfoster**
If the leading code does not end with a blank line and the student's code begins on the first line, or if the student's code does not end with a blank line and the trailing code begins on the first line, then the two files will overlap resulting in a syntax error with the code is run. While one can work around this problem by adding blank lines (good practice at the end of a file, less common at the beginning of a file), the autograder would be more resillient if the system ensured proper newlines. This is an attempt to add such code.

### CODE REVIEW

**Comment by jonatanschroeder** [comment]
I have been bitten by the same issue in the past as well, so I'd be in favour of this change. That said, I see that this can be a bit of a breaking change if an instructor relies on the current behaviour, though the same concern you have raised about the status quo (i.e., resiliency based on expectations of line breaks, or lack of them) would probably apply in that case if a student ends up adding a line break where they were not expected. So I am in favour of this change, but I'll wait for someone with higher authority to give the official approval. Also, I'd suggest that merging this change is done between terms to avoid possibly breaking current behaviour in the middle of the term, though I'd be fine if someone decides otherwise.

In essence, right now there is a way to ensure the line breaks are there, namely by ensuring that leading_code.py ends with a line break, and trailing_code.py starts with a line break (I typically start it with a comment for a similar reason). Removing that requirement (particularly for trailing_code.py) would avoid problems that may creep up in these questions.

**Comment by jgfoster** [comment]
@jonatanschroeder, Thanks for your comments. I actually ran into the problem when implementing your suggestion from Sept 22 on Slack about how to capture stdout!

**Comment by mwest1066** [comment]
I checked existing uses of `leading_code.py` and `trailing_code.py` and there are many examples in use that do not have a terminating newline, meaning that this PR would result in a change of behavior for these questions. That being said, all of the examples that I manually spot-checked looked like `import bumpy as np` or similar, so they are actually incorrect at the moment and would work better with this PR.

On balance, I think we should merge this PR, but I am slightly nervous about doing so in the middle of semester. Some options:
1. Wait until the December holiday break to merge it.
2. Just merge it now and hope/assume that no one is affected.
3. Email all courses using `leading_code.py`/`trailing_code.py` and warn them that in two weeks we are about to make this change.

I guess I lean towards (3). @nwalters512?

File: graders/python/python_autograder/pl_execute.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ def execute_code(
             str_student = pl_helpers.extract_ipynb_contents(f, ipynb_key)
         else:
             str_student = f.read()
-    str_student = str_leading + str_student + str_trailing
+    str_student = str_leading + "\n" + str_student + "\n" + str_trailing
```
Comment (mwest1066) [inline graders/python/python_autograder/pl_execute.py:80]: Would it be less surprising to users to only add the `\n` if the previous string was not already newline terminated? And should we also add a final newline on the very end (after `str_trailing`)?

**Review by mwest1066 — APPROVED** [review]

**Comment by jonatanschroeder** [comment]
> import bumpy as np

Never heard of [bumpy](https://pypi.org/project/bumpy/) before... 😝 

> Would it be less surprising to users to only add the `\n` if the previous string was not already newline terminated? 

I think that's a good idea for str_leading. I don't think that's necessary for str_student, but I wouldn't object to it.

> And should we also add a final newline on the very end (after `str_trailing`)?

Not sure if that's necessary, but I wouldn't object to it.

> On balance, I think we should merge this PR, but I am slightly nervous about doing so in the middle of semester. Some options:
> 
> 1. Wait until the December holiday break to merge it.
> 2. Just merge it now and hope/assume that no one is affected.
> 3. Email all courses using `leading_code.py`/`trailing_code.py` and warn them that in two weeks we are about to make this change.
> 
> I guess I lean towards (3). @nwalters512?

Would it be too disruptive/intensive to do (3) only for instructors of courses that have a `leading_code.py` that doesn't end with a line break? That said, I'm not overly concerned and wouldn't be opposed with (1) or (2).

**Comment by mwest1066** [comment]
> Never heard of [bumpy](https://pypi.org/project/bumpy/) before... 😝

lolol

> > 3. Email all courses using `leading_code.py`/`trailing_code.py` and warn them that in two weeks we are about to make this change.
> 
> Would it be too disruptive/intensive to do (3) only for instructors of courses that have a `leading_code.py` that doesn't end with a line break? That said, I'm not overly concerned and wouldn't be opposed with (1) or (2).

Oh yes, sorry, I did mean to only email folks who would actually be impacted by the change (as you say, if they have a code file without a terminal newline).

We're generally trying to be more cautious with deploying behavior changes. We've been burned a few times recently by deploying things that we didn't think were a big deal, but then it caused a lot of problems for some instructors when things "broke" for them. I think this change falls into that category, where it seems fine, but it's technically a breaking change, so we are probably better off being safe than sorry. All that being said, of course I want to discuss with @nwalters512 first.

**Comment by jgfoster** [comment]
I've merged the latest `master` and am willing to make edits. My only reluctance is that something more complicated would be riskier (in the sense of being error prone).

**Comment by nwalters512** [comment]
Tl;dr I did some work to validate that this wouldn't break anyone's questions, and I think we're safe.

I used the following one-liner to collect every `leading_code.py` from every question in production:

```sh
find courses -path "*/questions/**/leading_code.py" -name "leading_code.py" -type f -size +0c -exec sh -c 'if [ "$(tail -c1 "{}" | xxd -p)" != "0a" ]; then echo "=== {} ==="; cat "{}"; echo; fi' \;
```

This produced a file like the following:

```
=== courses/.../leading_code.py ===
import numpy as np

values = [0, np.pi / 2, np.pi, -np.pi / 2]
=== courses/.../leading_code.py ===

import numpy as np
from scipy.integrate import solve_ivp
=== courses/.../leading_code.py ===
Lab5="""
```

I then ran the following script against that file:

```py
import ast

files = {}

with open("data.txt") as f:
    filename = None
    for line in f:
        if line.startswith("==="):
            filename = line.strip()[4:-4]
            files[filename] = []
        else:
            files[filename].append(line)

for file in files:
    contents = "".join(files[file])
    try:
        ast.parse(contents)
    except Exception as exc:
        print(f"{file}: {exc}")
        print(contents)
        print()
```

My thinking is that if we can parse the leading code as a valid Python program, it's safe and wouldn't be impacted by the addition of a trailing newline.

This revealed two types of files that would potentially be impacted by this change:

```
courses/.../leading_code.py: expected an indented block after function definition on line 1 (<unknown>, line 1)
def answer():


courses/.../leading_code.py: unterminated triple-quoted string literal (detected at line 1) (<unknown>, line 1)
Lab5="""
```

For the first one, this might appear to cause problems, but I looked at the individual question and they're already manipulating the submitted answer in `parse()` to a) include a leading newline and b) insert indents on each line so that the student's answer is treated as a function body. This question wouldn't be impacted.

For the second one, they're ~abusing~ using the Python grader to grade a student's submitted HTML file by configuring their test as follows:

```py
student_code_file = "Lab5.html"
```

They then get access to the HTML via `Lab5`. This will also not be impacted by a trailing newline.

So, I think we're safe here, but I'd like either @mwest1066 or @jonatanschroeder to check my work (I already chatted with @eliotwrobson and he agrees that this should be fine).

**Comment by jonatanschroeder** [comment]
I agree that we're safe.

File: graders/python/python_autograder/pl_execute.py (hunk 1 above)
Comment (nwalters512) [reply graders/python/python_autograder/pl_execute.py:80]: I'm not entirely sure what difference it'd make, as this string is only ever used in memory via `exec(...)`, never written to disk. I think all we care about here is syntactic correctness, and one vs two newlines should never make a difference. Am I thinking about this correctly? cc @jgfoster @eliotwrobson

Comment (jgfoster) [reply graders/python/python_autograder/pl_execute.py:80]: My interest was syntactic correctness and two newlines should be fine.

Comment (mwest1066) [reply graders/python/python_autograder/pl_execute.py:80]: I'm fine with unconditionally adding another newline (so we might end up with two).

**Review by nwalters512 — APPROVED** [review]
